### PR TITLE
Windows X86: Fix pathological stackwalking case

### LINKED ISSUE #473: Stackwalking difference between rust-minidump and breakpad on Windows X86
We're seeing different stackwalking results between breakpad and rust-minidump on Windows X86, specifically in the case of FPO-type `STACK WIN` records.

Output from breakpad:
```
"frames": [
  {
    "status": "symbolicated",
    "original_index": 0,
    "instruction_addr": "0x77c6f911",
    "package": "C:\\Windows\\SysWOW64\\ntdll.dll",
    "symbol": "ZwWaitForSingleObject@12",
    "sym_addr": "0x77c6f8fc",
    "function": "ZwWaitForSingleObject@12",
    "lineno": 0,
    "trust": "context"
  },
  {
    "status": "symbolicated",
    "original_index": 1,
    "instruction_addr": "0x762414aa",
    "package": "C:\\Windows\\syswow64\\KERNELBASE.dll",
    "symbol": "WaitForSingleObjectEx@12",
    "sym_addr": "0x76241413",
    "function": "WaitForSingleObjectEx@12",
    "lineno": 0,
    "trust": "cfi"
  },
  […]
]
```

Output from rust-minidump:
```
"frames": [
  {
    "status": "symbolicated",
    "original_index": 0,
    "instruction_addr": "0x77c6f911",
    "package": "C:\\Windows\\SysWOW64\\ntdll.dll",
    "symbol": "ZwWaitForSingleObject@12",
    "sym_addr": "0x77c6f8fc",
    "function": "ZwWaitForSingleObject@12",
    "lineno": 0,
    "trust": "context"
  },
  {
    "status": "symbolicated",
    "original_index": 1,
    "instruction_addr": "0x77c6f910",
    "package": "C:\\Windows\\SysWOW64\\ntdll.dll",
    "symbol": "ZwWaitForSingleObject@12",
    "sym_addr": "0x77c6f8fc",
    "function": "ZwWaitForSingleObject@12",
    "lineno": 0,
    "trust": "cfi"
  },
  {
    "status": "symbolicated",
    "original_index": 2,
    "instruction_addr": "0x762414aa",
    "package": "C:\\Windows\\syswow64\\KERNELBASE.dll",
    "symbol": "WaitForSingleObjectEx@12",
    "sym_addr": "0x76241413",
    "function": "WaitForSingleObjectEx@12",
    "lineno": 0,
    "trust": "cfi"
  },
  […]
]
```
The rest of the stacktrace is the same, i.e., rust-minidump finds an additional frame immediately after the context. We suspect the reason lies somewhere in the fiddling that breakpad does with `.raSearchStart` starting [at this line](https://github.com/google/breakpad/blob/78f7ae495bc147e97a58e8158072fd35fdd99419/src/processor/stackwalker_x86.cc#L226). It's not clear to me personally why they scan for a return address in the middle of unwinding via CFI.

### PR DESCRIPTION
**Pull request by loewenheim**
This would fix #473. Compared to Breakpad's treatment of this case (see the linked issue), our check is more lax: they additionally check whether they're currently in a context frame, but we don't have this information in this function.

### CODE REVIEW

**Review by Gankra — APPROVED** [review]
Gross, but seems reasonable. Approved assuming you'll include a test.

If you think it's important to restrict this to the context frame, you can add that information to the FrameWalker trait, which basically just includes whatever this code needs.

**Comment by loewenheim** [comment]
> If you think it's important to restrict this to the context frame, you can add that information to the FrameWalker trait, which basically just includes whatever this code needs.

`FrameTrust` is defined in `minidump-processor`, which depends on `breakpad-symbols`. This would introduce a cyclic dependency.

**Comment by Gankra** [comment]
We already have the notion of a "grand_callee" so we could just expose `has_grand_callee(&self) -> bool` method as a proxy for "are we unwinding the context frame?"

**Comment by Gankra** [comment]
I'm not sure what anything would do with that extra information?

File: breakpad-symbols/src/sym_file/walker.rs [hunk 1]
```diff
@@ -987,9 +987,23 @@ pub fn walk_with_stack_win_fpo(info: &StackInfoWin, walker: &mut dyn FrameWalker
...
-        let eip_address = callee_esp + frame_size;
-        let caller_eip = walker.get_register_at_address(eip_address)?;
-        let caller_esp = callee_esp + frame_size + 4;
+        let mut eip_address = callee_esp + frame_size;
+        let mut caller_eip = walker.get_register_at_address(eip_address)?;
+
+        // Check whether
+        // 1. we are in a context frame. We approximate this by checking whether there's a grand callee.
+        // 2. the last return address was left on the stack after returning. This is the case if the caller's
+        //    eip (aka the return address) is the same the callee's eip.
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/walker.rs:996]:
```suggestion
        //    eip (aka the return address) is the same as the callee's eip.
```

**Comment by luser** [comment]
> I'm not sure what anything would do with that extra information?

The only thing that I think would be better would be `current_stack_depth == 1` as a much more direct "this is the context frame" signal than the `has_grand_callee` proposal. You could simplify this further and just directly provide an `is_context_frame: bool` argument.

**Comment by Gankra** [comment]
Haha, well, on the other hand I immediately was surprised by `current_stack_depth == 1` since I would have expected 0 indexing!

**Comment by loewenheim** [comment]
I added a unit test, do you think it's sufficient?

**Comment by Gankra** [comment]
This is top tier quality! Ready to land with a squash.

**Comment by loewenheim** [comment]
> This is top tier quality! Ready to land with a squash.

Aww, thanks!

**Pull request merged**